flow/connectors/elasticsearch/elasticsearch.go

```diff
@@ -0,0 +1,210 @@
+}
+
+func (esc *ElasticsearchConnector) SyncQRepRecords(ctx context.Context, config *protos.QRepConfig,
+	partition *protos.QRepPartition, stream *model.QRecordStream,
+) (int, error) {
+	startTime := time.Now()
+
+	schema := stream.Schema()
+
+	var bulkIndexHasError bool
+	var bulkIndexErrors []error
+	var bulkIndexMutex sync.Mutex
+	var docId string
+	numRecords := 0
+	bulkIndexerHasShutdown := false
+
+	// -1 means use UUID, >=0 means column in the record
+	upsertColIndex := -1
+	// only support single upsert column for now
+	if config.WriteMode.WriteType == protos.QRepWriteType_QREP_WRITE_MODE_UPSERT &&
+		len(config.WriteMode.UpsertKeyColumns) == 1 {
+		for i, field := range schema.Fields {
+			if config.WriteMode.UpsertKeyColumns[0] == field.Name {
+				upsertColIndex = i
+			}
+		}
+	}
+
+	esBulkIndexer, err := esutil.NewBulkIndexer(esutil.BulkIndexerConfig{
+		Index:  config.DestinationTableIdentifier,
+		Client: esc.client,
+		// parallelism comes from the workflow design itself, no need for this
+		NumWorkers:    1,
+		FlushInterval: 10 * time.Second,
+		// 1GB
+		FlushBytes: 1_000_000_000,
+	})
+	if err != nil {
+		esc.logger.Error("[es] failed to initialize bulk indexer", slog.Any("error", err))
+		return 0, fmt.Errorf("[es] failed to initialize bulk indexer: %w", err)
+	}
+	defer func() {
+		if !bulkIndexerHasShutdown {
+			err := esBulkIndexer.Close(context.Background())
+			if err != nil {
+				esc.logger.Error("[es] failed to close bulk indexer", slog.Any("error", err))
+			}
+		}
+	}()
+
+	for qRecordOrErr := range stream.Records {
+		if qRecordOrErr.Err != nil {
+			esc.logger.Error("[es] failed to get record from stream", slog.Any("error", qRecordOrErr.Err))
+			return 0, fmt.Errorf("[es] failed to get record from stream: %w", qRecordOrErr.Err)
+		}
+
+		qRecordJsonMap := make(map[string]any)
+
+		if upsertColIndex >= 0 {
+			docId = fmt.Sprintf("%v", qRecordOrErr.Record[upsertColIndex].Value())
+		} else {
+			docId = uuid.New().String()
+		}
+		for i, field := range schema.Fields {
+			switch field.Type {
+			// JSON is stored as a string, fix that
+			case qvalue.QValueKindJSON:
+				var jsonUnnest any
+				err := json.Unmarshal([]byte(qRecordOrErr.Record[i].Value().(string)), &jsonUnnest)
+				if err != nil {
+					esc.logger.Error("[es] failed to json.Unmarshal subfield", slog.Any("error", qRecordOrErr.Err))
+					return 0, fmt.Errorf("[es] failed to json.Unmarshal subfield: %w", qRecordOrErr.Err)
+				}
+				qRecordJsonMap[field.Name] = jsonUnnest
+			default:
+				qRecordJsonMap[field.Name] = qRecordOrErr.Record[i].Value()
+			}
+		}
+		qRecordJsonBytes, err := json.Marshal(qRecordJsonMap)
+		if err != nil {
+			esc.logger.Error("[es] failed to json.Marshal record", slog.Any("error", err))
+			return 0, fmt.Errorf("[es] failed to json.Marshal record: %w", err)
+		}
+
+		esBulkIndexer.Add(ctx, esutil.BulkIndexerItem{
+			Action:     "index",
+			DocumentID: docId,
+			Body:       bytes.NewReader(qRecordJsonBytes),
```

Comment:
oh we could straight up pass an actual reader instead of `bytes.NewReader` here & stream the above json encoding (convert appends into `Write` calls) but would have to spin up a goroutine so probably not worth

Reply:
doesn't interface well with QRep parallelism implemented across multiple workflows currently